src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/SelfHosted/SelfHostedSyncSponsorshipsCommand.cs

```diff
@@ -0,0 +1,129 @@
+            var syncedSponsorships = new List<OrganizationSponsorshipData>();
+
+            foreach (var orgSponsorshipsBatch in CoreHelpers.Batch(organizationSponsorships, 1000))
+            {
+                var response = await SendAsync<OrganizationSponsorshipSyncRequestModel, OrganizationSponsorshipSyncResponseModel>(HttpMethod.Post, "organizationSponsorships/sync", new OrganizationSponsorshipSyncRequestModel
+                {
+                    BillingSyncKey = billingSyncKey.Config,
+                    SponsoringOrganizationCloudId = cloudOrganizationId,
+                    SponsorshipsBatch = orgSponsorshipsBatch.Select(s => new OrganizationSponsorshipRequestModel
+                    {
+                        SponsoringOrganizationUserId = s.SponsoringOrganizationUserId,
+                        FriendlyName = s.FriendlyName,
+                        OfferedToEmail = s.OfferedToEmail,
+                        PlanSponsorshipType = s.PlanSponsorshipType.GetValueOrDefault(),
+                        ValidUntil = s.ValidUntil,
+                        ToDelete = s.ToDelete
+                    })
+                });
+
+                if (response == null)
+                {
+                    throw new BadRequestException("Organization sync failed");
+                }
+
+                syncedSponsorships.AddRange(response.ToOrganizationSponsorshipSync().SponsorshipsBatch);
+            }
+
+            var organizationSponsorshipsDict = organizationSponsorships.ToDictionary(i => i.SponsoringOrganizationUserId);
+            var sponsorshipsToDelete = syncedSponsorships.Where(s => s.CloudSponsorshipRemoved).Select(i => organizationSponsorshipsDict[i.SponsoringOrganizationUserId].Id);
+            var sponsorshipsToUpsert = syncedSponsorships.Where(s => !s.CloudSponsorshipRemoved).Select(i =>
+            {
+                var existingSponsorship = organizationSponsorshipsDict[i.SponsoringOrganizationUserId];
+                if (existingSponsorship != null)
+                {
+                    existingSponsorship.LastSyncDate = i.LastSyncDate;
+                    existingSponsorship.ValidUntil = i.ValidUntil;
+                    existingSponsorship.ToDelete = i.ToDelete;
+
+                }
+                else
+                {
+                    existingSponsorship = new OrganizationSponsorship
+                    {
+                        SponsoringOrganizationId = organizationId,
+                        SponsoringOrganizationUserId = i.SponsoringOrganizationUserId,
+                        FriendlyName = i.FriendlyName,
+                        OfferedToEmail = i.OfferedToEmail,
+                        PlanSponsorshipType = i.PlanSponsorshipType,
+                        LastSyncDate = i.LastSyncDate,
+                        ValidUntil = i.ValidUntil,
+                        ToDelete = i.ToDelete
+                    };
+                }
```

Comment:
This is never intended to happen, right?

If so, please add a comment that this shouldn't occur